crates/executor/src/syscalls/eip2935.rs

```diff
@@ -0,0 +1,183 @@
+        ))
+        .build();
+
+    // initialize a block from the env, because the pre block call needs the block itself
+    apply_block_hash_contract_call(
+        config,
+        payload.payload_attributes.timestamp,
+        block_number,
+        parent_block_hash,
+        &mut evm_pre_block,
+    )
+}
+
+/// Apply the EIP-4788 pre-block beacon root contract call to a given EVM instance.
+fn apply_block_hash_contract_call<F, H>(
+    config: &RollupConfig,
+    timestamp: u64,
+    block_number: u64,
+    parent_block_hash: B256,
+    evm: &mut Evm<'_, (), &mut State<&mut TrieDB<F, H>>>,
+) -> ExecutorResult<()>
+where
+    F: TrieDBProvider,
+    H: TrieHinter,
+{
+    if !config.is_isthmus_active(timestamp) {
+        return Ok(());
+    }
+
+    // If the block number is zero (genesis block) then no system
+    // transaction may occur as per EIP-2935.
+    if block_number == 0 {
+        return Ok(())
+    }
+
+    // Get the previous environment
+    let previous_env = Box::new(evm.context.evm.env().clone());
+
+    // modify env for pre block call
+    fill_tx_env_with_block_hash_contract_call(&mut evm.context.evm.env, parent_block_hash);
+
+    let mut state = match evm.transact() {
+        Ok(res) => res.state,
+        Err(e) => {
+            evm.context.evm.env = previous_env;
+            return Err(ExecutorError::ExecutionError(e));
+        }
+    };
+
+    // NOTE: Revm currently marks these accounts as "touched" when we do the above transact calls,
+    // and includes them in the result.
+    //
+    // There should be no state changes to these addresses anyways as a result of this system call,
+    // so we can just remove them from the state returned.
+    state.remove(&alloy_eips::eip4788::SYSTEM_ADDRESS);
+    state.remove(&evm.block().coinbase);
+
+    evm.context.evm.db.commit(state);
+
+    // re-set the previous env
+    evm.context.evm.env = previous_env;
+
+    Ok(())
+}
+
+/// Fill transaction environment with the EIP-2935 system contract message data.
+///
+/// This requirements for the beacon root contract call defined by
+/// [EIP-2935](https://eips.ethereum.org/EIPS/eip-2935) are:
+fn fill_tx_env_with_block_hash_contract_call(env: &mut Env, parent_block_hash: B256) {
+    fill_tx_env_with_system_contract_call(
+        env,
+        alloy_eips::eip4788::SYSTEM_ADDRESS,
+        alloy_eips::eip2935::HISTORY_STORAGE_ADDRESS,
+        parent_block_hash.into(),
+    );
+}
+
+/// Fill transaction environment with the system caller and the system contract address and message
+/// data.
+///
+/// This is a system operation and therefore:
+///  * the call must execute to completion
+///  * the call does not count against the block’s gas limit
+///  * the call does not follow the EIP-1559 burn semantics - no value should be transferred as part
+///    of the call
+///  * if no code exists at the provided address, the call will fail silently
+fn fill_tx_env_with_system_contract_call(
```

Comment:
can this be shared with EIP-4788 + EIP-7251? This code should already exist in `eip4788.rs`.